modules/globebrowsing/src/screenspacevideorenderable.cpp

```diff
@@ -0,0 +1,359 @@
+    av_image_fill_arrays(
+        _glFrame->data,
+        _glFrame->linesize,
+        internalBuffer,
+        AV_PIX_FMT_RGB24,
+        _codecContext->width,
+        _codecContext->height,
+        1
+    );
+
+    // Update times
+    _lastFrameTime = std::chrono::system_clock::now();
+
+    // Create the texture
+    _texture = std::make_unique<ghoul::opengl::Texture>(
+        glm::uvec3(_codecContext->width, _codecContext->height, 1),
+        GL_TEXTURE_2D,
+        ghoul::opengl::Texture::Format::RGB,
+        GL_RGB,
+        GL_UNSIGNED_BYTE,
+        ghoul::opengl::Texture::FilterMode::Linear,
+        ghoul::opengl::Texture::WrappingMode::Repeat,
+        ghoul::opengl::Texture::AllocateData::No,
+        ghoul::opengl::Texture::TakeOwnership::No
+    );
+
+    _conversionContext = sws_getContext(
+        _codecContext->width,
+        _codecContext->height,
+        _codecContext->pix_fmt,
+        _codecContext->width,
+        _codecContext->height,
+        AV_PIX_FMT_RGB24,
+        SWS_BICUBIC,
+        nullptr,
+        nullptr,
+        nullptr
+    );
+
+    readFrame(); 
+    _isInitialized = true;
+    _objectSize = { _codecContext->width, _codecContext->height };
+
+    return true;
+}
+
+bool ScreenSpaceVideoRenderable::deinitializeGL() {
+    avformat_close_input(&_formatContext);
+    av_free(_avFrame);
+    av_free(_glFrame);
+    av_free(_packet);
+    avformat_free_context(_formatContext);
+
+    ScreenSpaceRenderable::deinitializeGL();
+    return true;
+}
+
+glm::mat4 ScreenSpaceVideoRenderable::localRotationMatrix() {
+    // Ffmpeg reads the texture differently than OpenGL. Flip the texture to make it 
+    // appear correctly
+    glm::mat4 flipTexture = glm::mat4(glm::quat(glm::vec3(glm::pi<float>(), 0.f, 0.f)));
+    return ScreenSpaceRenderable::localRotationMatrix() * flipTexture;
+}
+
+void ScreenSpaceVideoRenderable::render() {
+    ScreenSpaceRenderable::render();
+}
+
+void ScreenSpaceVideoRenderable::update() {
+
+    ZoneScoped
+
+    if (!_isInitialized || !_isPlaying) {
+        return;
+    }
+    // Check if it is time for a new frame
+    std::chrono::system_clock::time_point now = std::chrono::system_clock::now();
+    std::chrono::system_clock::duration diff = now - _lastFrameTime;
+
+    const bool hasNewFrame = diff > _frameTime;
+
+    if (!hasNewFrame) {
+        return;
+    }
+
+    readFrame();
+
+    // Update times
+    if (_frameTime.count() <= 0) {
+        // Calculate frame time
+        double sPerFrame = av_q2d(_codecContext->time_base) * _codecContext->ticks_per_frame;
+        int msPerFrame = static_cast<int>(sPerFrame * 1000);
+        _frameTime = std::chrono::milliseconds(msPerFrame);
+    }
+    _lastFrameTime = now;
+
+    ScreenSpaceRenderable::update();
+}
+
+void ScreenSpaceVideoRenderable::readFrame() {
+    while (true) {
+        int result = av_read_frame(_formatContext, _packet);
+        if (result < 0) {
```

Comment:
what does this mean?  Here I would benefit from a comment or descriptive variable name `bool somethingMeaningful = result < 0;` Right now I don't get what this check is for.

Reply:
Maybe just a comment for what `result < 0` menas here would be enough. Then I would understand all the places it happens below as well